File: tests/typesafety/typescript/typed_pipeline_misuse.ts
// expect: fails

type RawOrder = { readonly item: string; readonly quantityText: string };
type Order = { readonly item: string; readonly quantity: number };
type Priced = { readonly item: string; readonly totalCents: number };

function price(order: Order, unitCents: number): Priced {
  return { item: order.item, totalCents: order.quantity * unitCents };
}

export const quoted = price({ item: "saddle", quantityText: "2" }, 250);

File: tests/typesafety/python/typed_pipeline_misuse.py
# expect: fails
# title: A RawOrder handed straight to price, rejected by the checker
# misuse-of: typed_pipeline
from dataclasses import dataclass


@dataclass(frozen=True)
class RawOrder:
    item: str
    quantity_text: str


@dataclass(frozen=True)
class Order:
    item: str
    quantity: int


@dataclass(frozen=True)
class Priced:
    item: str
    total_cents: int


def price(order: Order, unit_cents: int) -> Priced:
    return Priced(item=order.item, total_cents=order.quantity * unit_cents)


quoted = price(RawOrder(item="saddle", quantity_text="2"), 250)
